rfcs/45-stageless.md

```diff
@@ -0,0 +1,823 @@
+This will become increasingly important as more complex graph-builder APIs are added.
+
+**In summary:**
+
+- array syntax: optimal but literally impossible
+- tuple syntax: magic but pretty
+- builder syntax: way too verbose for a pattern that's supposed to increase convenience
+- `vec!`-style macro: not any less magic than tuple syntax, not as pretty
+
+**Conclusion:** tuple syntax, powered by builder syntax under the hood that users can use instead when it's more convenient.
+
+Below, we examine the the options by example.
+
+#### Arrays
+
+```rust
+.add_systems([compute_attack, 
+                   compute_defense,
+                   check_for_crits,
+                   compute_damage,
+                   deal_damage,
+                   check_for_death]
+                   .chain()
+                   .label(GameLabel::Combat))
+
+.add_systems([run, jump].after(CoreLabel::Input))
+```
+
+Very pretty. Doesn't work because types are heterogenous though.
+
+#### Tuples
+
+```rust
+.add_systems((compute_attack, 
+              compute_defense,
+              check_for_crits,
+              compute_damage,
+              deal_damage,
+              check_for_death)
+              .chain()
+              .label(GameLabel::Combat)
+            )
+
+.add_systems((run, jump).after(CoreLabel::Input))
+```
+
+Also pretty! Mildly cursed and definitely unidiomatic, but we use this flavor of cursed type system magic already for bundles.
+
+Requires another invocation of the `all_tuples` macro to generate impls.
+
+#### Builder pattern
+
+This was the strategy we used for `SystemSet`.
+
+```rust
+.add_systems(SystemGroup:::new()
+               .with(compute_attack), 
+               .with(compute_defense),
+               .with(check_for_crits),
+               .with(compute_damage),
+               .with(deal_damage),
+               .with(check_for_death)
+               .chain()
+               .label(GameLabel::Combat)
+            )
+
+.add_systems(SystemGroup::new().with(run).with(jump).after(CoreLabel::Input))
+```
+
+Oof. Very verbose and cluttered, even with a change from `with_system` to `with`.
+Placement of brackets can be confusing, and there's no clear seperation between systems and their shared config.
+
+This is particularly painful for small groups, and is not significantly more ergonomic than just adding the systems individually.
+
+#### Vec-style macro
+
+This strategy is used for the `vec![1,2,3]` macro in the standard library.
+
+```rust
+.add_systems(systems![
+      compute_attack, 
+      compute_defense,
+      check_for_crits,
+      compute_damage,
+      deal_damage,
+      check_for_death
+   ]
+   .chain()
+   .label(GameLabel::Combat)
+)
+
+.add_systems(systems![run, jump].after(CoreLabel::Input))
+```
+
+Reasonably pretty, but still more verbose than the tuple option.
+Requires boilerplate invocation of a macro every single time these methods are used.
+Not any less magic than the tuples.

```

Comment:
I think this misses the fact that the tuple way of doing things will just spit out a horrible error after a certain length whereas `systems![...]` can work for any length

Reply:
Hmm. That's a pretty compelling argument. I think I like the macro better then on the whole.

Reply:
I find it a big plus point for Bevy that it currently requires very few macros in user code. As long as there is some way of grouping (like adding bundles to bundles, or grouping system params), I would prefer the tuple way.

Here, you can always add more systems to the set later or just use the builder functions, right?